pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/channel/MailboxContentStreamObserver.java

```diff
@@ -59,13 +63,32 @@ private static Mailbox.MailboxContent createErrorContent(Throwable e)
   private final boolean _isEnabledFeedback;
 
   private final AtomicBoolean _isCompleted = new AtomicBoolean(false);
-  private final ArrayBlockingQueue<Mailbox.MailboxContent> _receivingBuffer;
-
-  ReadWriteLock _errorLock = new ReentrantReadWriteLock();
-  private Mailbox.MailboxContent _errorContent = null; // Guarded by _errorLock.
+  private final BlockingQueue<Mailbox.MailboxContent> _receivingBuffer;
+
+  private ReadWriteLock _bufferSizeLock = new ReentrantReadWriteLock();
+  @GuardedBy("bufferSizeLock")
+  private int _maxBufferSize = 0;
+  private ReadWriteLock _errorLock = new ReentrantReadWriteLock();
+  @GuardedBy("_errorLock")
+  private Mailbox.MailboxContent _errorContent = null;
   private StringMailboxIdentifier _mailboxId;
   private Consumer<MailboxIdentifier> _gotMailCallback;
 
+  private void updateMaxBufferSize() {
+    _bufferSizeLock.writeLock().lock();
+    _maxBufferSize = max(_maxBufferSize, _receivingBuffer.size());
+    _bufferSizeLock.writeLock().unlock();
+  }
+
+  private int getMaxBufferSize() {
+    try {
+      _bufferSizeLock.readLock().lock();
+      return _maxBufferSize;
+    } finally {
+      _bufferSizeLock.readLock().unlock();
+    }
+  }
```

Comment:
since this is only for printing debug purpose. do we need to guard them in the lock?

Reply:
yes. it is unsafe even for debugging purpose